synapse/storage/roommember.py

```diff
@@ -501,40 +501,34 @@ def _get_joined_users_from_context(self, room_id, state_group, current_state_ids
 
         defer.returnValue(users_in_room)
 
-    def is_host_joined(self, room_id, host, state_group, state_ids):
-        if not state_group:
-            # If state_group is None it means it has yet to be assigned a
-            # state group, i.e. we need to make sure that calls with a state_group
-            # of None don't hit previous cached calls with a None state_group.
-            # To do this we set the state_group to a new object as object() != object()
-            state_group = object()
+    @defer.inlineCallbacks
+    def is_host_joined(self, room_id, host):
+        if '%' in host or '_' in host:
+            raise Exception("Invalid host name")
+
+        sql = """
+            SELECT state_key FROM current_state_events
+            INNER JOIN room_memberships USING (room_id, event_id)
+            WHERE membership = 'join' AND room_id = ? AND state_key LIKE ?
+            LIMIT 1
```

Comment:
It might help the query optimiser if you threw in a `type = 'm.room.member'` on the where clause so that it can better use the `current_state_events_room_id_type_state_key_key index`.

I think you might want to remove the "`room_id`" from the join since we want it to force it to join using the `room_memberships_event_id_key` index.

On matrix.org:
```
matrix=# explain SELECT state_key FROM current_state_events 
  INNER JOIN room_memberships USING (room_id, event_id) 
  WHERE membership = 'join'
  AND room_id = '!cURbafjkfsMDVwdRDQ:matrix.org' 
  AND state_key LIKE '%:matrix.org';

                                                        QUERY PLAN                                                        
--------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=898.53..95521.12 rows=11 width=18)
   ->  Bitmap Heap Scan on current_state_events  (cost=897.97..32646.62 rows=7500 width=83)
         Recheck Cond: (room_id = '!cURbafjkfsMDVwdRDQ:matrix.org'::text)
         Filter: (state_key ~~ '%:matrix.org'::text)
         ->  Bitmap Index Scan on current_state_events_room_id_type_state_key_key  (cost=0.00..896.09 rows=12205 width=0)
               Index Cond: (room_id = '!cURbafjkfsMDVwdRDQ:matrix.org'::text)
   ->  Index Scan using room_memberships_event_id_key on room_memberships  (cost=0.56..8.37 rows=1 width=65)
         Index Cond: (event_id = current_state_events.event_id)
         Filter: ((room_id = '!cURbafjkfsMDVwdRDQ:matrix.org'::text) AND (membership = 'join'::text))
(9 rows)

matrix=# explain SELECT state_key FROM current_state_events
  INNER JOIN room_memberships USING (event_id)
  WHERE membership = 'join'
  AND current_state_events.room_id = '!cURbafjkfsMDVwdRDQ:matrix.org'
  AND state_key LIKE '%:matrix.org'
  AND type='m.room.member';
                                                       QUERY PLAN                                                        
-------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=591.87..63361.07 rows=2660 width=18)
   ->  Bitmap Heap Scan on current_state_events  (cost=591.31..23057.21 rows=4766 width=52)
         Recheck Cond: ((room_id = '!cURbafjkfsMDVwdRDQ:matrix.org'::text) AND (type = 'm.room.member'::text))
         Filter: (state_key ~~ '%:matrix.org'::text)
         ->  Bitmap Index Scan on current_state_events_room_id_type_state_key_key  (cost=0.00..590.11 rows=7756 width=0)
               Index Cond: ((room_id = '!cURbafjkfsMDVwdRDQ:matrix.org'::text) AND (type = 'm.room.member'::text))
   ->  Index Scan using room_memberships_event_id_key on room_memberships  (cost=0.56..8.45 rows=1 width=34)
         Index Cond: (event_id = current_state_events.event_id)
         Filter: (membership = 'join'::text)

```